src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php

```diff
@@ -124,7 +122,7 @@ public function configureOptions(OptionsResolver $resolver)
             if (null === $options['choices']) {
                 // We consider two query builders with an equal SQL string and
                 // equal parameters to be equal
-                $qbParts = $options['query_builder']
+                $qbParts = $options['query_builder'] instanceof \Doctrine\ORM\QueryBuilder
                     ? array(
                         $options['query_builder']->getQuery()->getSQL(),
                         $options['query_builder']->getParameters()->toArray(),
```

Comment:
This logic to compare the query builder is specific to the ORM, so it must go in the child class, with a way for other Doctrine implementations (in MongoDBBundle for instance) to provide their own hashing for their query builders.

and in case the query builder cannot be compared (MongoDBBundle not implementing the method yet), the cache should be avoided instead of using the wrong cache.